packages/ckeditor5-ui/src/dialog/dialog.ts

```diff
@@ -0,0 +1,129 @@
+/**
+ * @license Copyright (c) 2003-2023, CKSource Holding sp. z o.o. All rights reserved.
+ * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
+ */
+
+/**
+ * @module ui/dialog/dialog
+ */
+
+import { type Editor, Plugin } from '@ckeditor/ckeditor5-core';
+import DialogView, { type DialogViewCloseEvent } from './dialogview';
+
+import '../../theme/components/dialog/dialog.css';
+import View from '../view';
+import type { DialogActionButtonDefinition } from './dialogactionsview';
+
+/**
+ * TODO
+ */
+export default class Dialog extends Plugin {
+	/**
+	 * TODO
+	 */
+	public readonly view: DialogView;
+
+	/**
+	 * TODO
+	 */
+	private _onHide: ( ( dialog: Dialog ) => void ) | undefined;
+
+	/**
+	 * @inheritDoc
+	 */
+	public static get pluginName() {
+		return 'Dialog' as const;
+	}
+
+	/**
+	 * @inheritDoc
+	 */
+	constructor( editor: Editor ) {
+		super( editor );
+
+		this.view = new DialogView( editor.locale );
+
+		this.view.on<DialogViewCloseEvent>( 'close', () => {
+			this.hide();
+		} );
+
+		editor.ui.view.body.add( this.view );
+		editor.ui.focusTracker.add( this.view.element! );
+	}
+
+	/**
+	 * TODO
+	 */
+	public show( {
```

Comment:
After few days I still think it would be great to have an event we could listen to. Decorated method (or something that acts like it) would be best as you could possibly change the incoming parameters.

Both show and hide should fire events.

Additionally, the passed `onShow` and `onHide` should also be included in that event. I am still not sure whether we will want `onShow` and `onHide` happen before or after the regular code defined in `show()` and `hide()`. For now let's stick to what we have now. I expect that `onShow` will not be popular (I'd even say it is not needed) as you can do same stuff just before firing `dialog.show()`. As for `onHide` I assume it won't matter.

If `show()` and `hide()` are decorated (or if they fire events directly), then the outside integrator will be able to add callbacks on `'high'` or `'low'` priority, having more flexibility. Additionally, they will be able to change the parameters.

But, if there's a dialog opened, I'd like `hide` event to be called first and `show` event to be called then. And this prevents us from decorating `show()` as it will fire the events in a different order...

So maybe it should be like this:

```javascript
show( params ) {
	if ( dialog opened ) {
		this.hide();
	}
	
	this.fire( 'show', ... ); // Pass `params` here so it works like a decorated method.
}

hide() {
	this.fire( 'hide' );
}
```

The regular code from these methods will be called with default priority.

```javascript
this.on( 'show', ( evt, args ) => {
	this._show( ...args );
} );
```

We could fire `onShow` and `onHide` as an callback to `show` and `hide` events too:

```javascript
this.on( 'show', ( evt, args ) => {
	args[ 0 ].onShow( ... ); // Not sure if this is correct, writing from top of my head.
}, { priority: 'low' } );
```

This will give extra flexibility so you will be able to add custom callback from outside to be called between `_show()` and `onShow`.